File: parse_cobertura.php
#!/usr/bin/env php
<?php
/**
 * parse_cobertura.php
 * Parses a Cobertura XML coverage report and outputs a Markdown summary.
 */
if ($argc !== 3) {
    fwrite(STDERR, "Usage: parse_cobertura.php <coverage.xml> <output.md>\n");
    exit(1);
}
$xmlPath = $argv[1];
$outputPath = $argv[2];
if (!file_exists($xmlPath)) {
    fwrite(STDERR, "Error: XML report file not found at '$xmlPath'\n");
    exit(1);
}
$xml = simplexml_load_file($xmlPath);
if ($xml === false) {
    fwrite(STDERR, "Error: Failed to parse XML file\n");
    exit(1);
}
// Overall coverage
$lineRate = ((float)($xml['line-rate'] ?? 0)) * 100;
$branchRate = ((float)($xml['branch-rate'] ?? 0)) * 100;
// Per-file classes coverage
$classes = [];
foreach ($xml->packages->package as $package) {
    foreach ($package->classes->class as $class) {
        $filename = (string)$class['filename'];
        $lr = ((float)($class['line-rate'] ?? 0)) * 100;
        $br = ((float)($class['branch-rate'] ?? 0)) * 100;
        $classes[] = ['filename' => $filename, 'line' => $lr, 'branch' => $br];
    }
}
// Sort classes by filename
usort($classes, function($a, $b) { return strcmp($a['filename'], $b['filename']); });
// Generate Markdown
$md = [];
$md[] = '# Coverage Report';
$md[] = '';
$md[] = "**Overall Line Coverage:** " . number_format($lineRate, 2) . "%  ";
$md[] = "**Overall Branch Coverage:** " . number_format($branchRate, 2) . "%  ";
$md[] = '';
$md[] = '## File Coverage  ';
$md[] = '';
$md[] = '| File | Line Coverage | Branch Coverage |';
$md[] = '| ---- | ------------- | --------------- |';
foreach ($classes as $c) {
    $md[] = sprintf(
        '| %s | %.2f%% | %.2f%% |',
        $c['filename'],
        $c['line'],
        $c['branch']
    );
}
// Write to output file
file_put_contents($outputPath, implode("\n", $md) . "\n");
exit(0);
?>

File: parse_cobertura.go
package main

import (
	"bufio"
	"encoding/xml"
	"flag"
	"fmt"
	"os"
	"slices"
	"sort"
	"strings"
)

// CoverageXML represents the root cobertura XML element
type CoverageXML struct {
	LineRate   float64      `xml:"line-rate,attr"`
	BranchRate float64      `xml:"branch-rate,attr"`
	Packages   []PackageXML `xml:"packages>package"`
}

// PackageXML represents a cobertura package element
type PackageXML struct {
	Classes []ClassXML `xml:"classes>class"`
}

// ClassXML represents a cobertura class element
type ClassXML struct {
	Filename   string  `xml:"filename,attr"`
	LineRate   float64 `xml:"line-rate,attr"`
	BranchRate float64 `xml:"branch-rate,attr"`
}

// FileCoverage holds coverage data for a single file
type FileCoverage struct {
	Filename   string
	LineRate   float64
	BranchRate float64
}

func generateBadge(coverage float64) string {
	if coverage >= 75 {
		return "green"
	}

	if coverage > 70 && coverage < 75 {
		return "yellow"
	}
	return "red"
}

func main() {
	inputPath := flag.String("input", "input.xml", "Path to Cobertura XML report file")
	changeListPath := flag.String("changeList", "i", "Path to change list file")
	outputPath := flag.String("output", "coverage.md", "Path to output Markdown file")
	flag.Parse()

	data, err := os.ReadFile(*inputPath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading XML file '%s': %v\n", *inputPath, err)
		os.Exit(1)
	}

	changeListFile, err := os.Open(*changeListPath)

	var changedFiles []string
	scanner := bufio.NewScanner(changeListFile)
	for scanner.Scan() {
		changedFiles = append(changedFiles, strings.TrimPrefix(scanner.Text(), "api/app/"))
	}

	var cov CoverageXML
	if err := xml.Unmarshal(data, &cov); err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing XML: %v\n", err)
		os.Exit(1)
	}

	var changedFilesCoverage []FileCoverage

	var avgCoverage float64
	// Collect file coverage
	var files []FileCoverage
	for _, pkg := range cov.Packages {
		for _, cls := range pkg.Classes {
			files = append(files, FileCoverage{
				Filename:   cls.Filename,
				LineRate:   cls.LineRate,
				BranchRate: cls.BranchRate,
			})

			if len(changedFiles) < 1 {
				continue
			}

			if slices.Contains(changedFiles, cls.Filename) {
				changedFilesCoverage = append(changedFilesCoverage, FileCoverage{
					Filename:   cls.Filename,
					LineRate:   cls.LineRate,
					BranchRate: cls.BranchRate,
				})
				avgCoverage += cls.LineRate
			}
		}
	}

	// Sort by filename
	sort.Slice(files, func(i, j int) bool {
		return files[i].Filename < files[j].Filename
	})

	// Build Markdown report
	var b strings.Builder
	b.WriteString("# Coverage Report\n\n")

	b.WriteString("\n")
	b.WriteString("\n")

	if len(changedFilesCoverage) > 0 {

		b.WriteString(fmt.Sprintf("**Changed Files Line Coverage:** %.2f%%  \n", avgCoverage/float64(len(changedFilesCoverage))*100))
		b.WriteString("\n")
		b.WriteString("\n")
		b.WriteString("![badge](https://img.shields.io/badge/Coverage-" + fmt.Sprintf("%f", avgCoverage/float64(len(changedFilesCoverage))*100) + "%25-" + generateBadge(avgCoverage/float64(len(changedFilesCoverage))*100) + ")")
		b.WriteString("\n")
		b.WriteString("\n")
		b.WriteString("| File | Line Coverage | \n")
		b.WriteString("| ---- | ------------- | \n")
		for _, f := range changedFilesCoverage {
			b.WriteString(fmt.Sprintf("| %s | %.2f%% |\n", f.Filename, f.LineRate*100))
		}
	}
	b.WriteString("\n")
	b.WriteString("\n")
	b.WriteString(fmt.Sprintf("**Overall Line Coverage:** %.2f%%  \n", cov.LineRate*100))
	b.WriteString("\n")
	b.WriteString("![badge](https://img.shields.io/badge/Coverage-" + fmt.Sprintf("%f", cov.LineRate*100) + "%25-" + generateBadge(cov.LineRate*100) + ")")
	b.WriteString("\n")
	b.WriteString("<details>")
	b.WriteString("\n")
	b.WriteString("<summary>File Coverage</summary>\n")
	b.WriteString("\n")
	b.WriteString("| File | Line Coverage | \n")
	b.WriteString("| ---- | ------------- | \n")
	for _, f := range files {
		b.WriteString(fmt.Sprintf("| %s | %.2f%% |\n", f.Filename, f.LineRate*100))
	}

	b.WriteString("</details>")

	// Write to output file
	if err := os.WriteFile(*outputPath, []byte(b.String()), 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing output file '%s': %v\n", *outputPath, err)
		os.Exit(1)
	}
}
